ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/datetime/ParseDate.java

```diff
@@ -73,8 +75,18 @@ public Date parseDate(
         throw new KsqlFunctionException("Date format contains time field.");
       }
 
-      return new Date(
-          TimeUnit.DAYS.toMillis(LocalDate.from(ta).toEpochDay()));
+      if (ta.isSupported(ChronoField.EPOCH_DAY)) {
+        return new Date(
+            TimeUnit.DAYS.toMillis(LocalDate.from(ta).toEpochDay()));
+      } else if (ta.isSupported(ChronoField.YEAR) && ta.isSupported(ChronoField.MONTH_OF_YEAR)) {
+        return new Date(
+            TimeUnit.DAYS.toMillis(YearMonth.from(ta).atDay(1).toEpochDay()));
+      } else if (ta.isSupported(ChronoField.YEAR) && !ta.isSupported(ChronoField.DAY_OF_MONTH)) {
+        return new Date(
+            TimeUnit.DAYS.toMillis(Year.from(ta).atDay(1).toEpochDay()));
+      } else {
+        throw new KsqlFunctionException("Cannot create DATE out of the fields provided.");
+      }
```

Comment:
So we are limited to use specific string formats. Seems not very flexible. Is there another library that allows us to use any string format we want?

Reply:
Swapped it to `SimpleDateFormat`, it does what we want. The difference is that it does even more partial parsing than we promised - it parses `PARSE_DATE("2021-04", "yyyy-dd")` to `2021-01-04`.

Reply:
Actually `SimpleDateFormat` isn't thread safe. `FastDateFormat` from `apache-commons`'s `lang3` library works though.